packages/block-library/src/query-title/edit.js

```diff
@@ -12,43 +12,91 @@ import {
 	BlockControls,
 	useBlockProps,
 	Warning,
+	RichText,
 } from '@wordpress/block-editor';
-import { __ } from '@wordpress/i18n';
+import { __, _x } from '@wordpress/i18n';
+import { useEffect } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import HeadingLevelDropdown from '../heading/heading-level-dropdown';
 
-const SUPPORTED_TYPES = [ 'archive' ];
+const SUPPORTED_TEMPLATES = [ 'archive', 'search', '404' ];
 
 export default function QueryTitleEdit( {
-	attributes: { type, level, textAlign },
+	attributes: { content, level, textAlign },
 	setAttributes,
+	context: { templateSlug },
 } ) {
 	const TagName = `h${ level }`;
 	const blockProps = useBlockProps( {
-		className: classnames( {
+		className: classnames( 'wp-block-query-title__placeholder', {
 			[ `has-text-align-${ textAlign }` ]: textAlign,
-			'wp-block-query-title__placeholder': type === 'archive',
 		} ),
 	} );
-	// The plan is to augment this block with more
-	// block variations like `Search Title`.
-	if ( ! SUPPORTED_TYPES.includes( type ) ) {
+	let titleElement;
+
+	// Infer title content from template slug context
+	// Defaults to content attribute prop
+	switch ( templateSlug ) {
+		case 'archive':
+			// translators: Title for archive template.
+			content = _x( 'Archive title', 'archive template title' );
+			titleElement = <TagName { ...blockProps }>{ content }</TagName>;
+			break;
+		case 'search':
+			// translators: Title for search template with dynamic content placeholders.
+			const placeholderContent = _x(
+				'Search results for "%search%"',
+				'search template title'
+			);
+
+			// Use placeholder content if new Query Title block is added
+			if ( content === '' ) {
+				content = placeholderContent;
+			}
+
+			titleElement = (
+				<div { ...blockProps }>
+					<RichText
+						tagName={ TagName }
+						value={ content }
+						placeholder={ placeholderContent }
+						allowedFormats={ [ 'core/bold', 'core/italic' ] }
+						onChange={ ( newContent ) =>
+							setAttributes( { content: newContent } )
+						}
+						disableLineBreaks={ true }
+					/>
+				</div>
+			);
+			break;
+		case '404':
+			// translators: Title for 404 template.
+			content = _x( 'Nothing found', '404 template title' );
+			titleElement = <TagName { ...blockProps }>{ content }</TagName>;
+			break;
+		default:
+			titleElement = <TagName { ...blockProps }>{ content }</TagName>;
+			break;
+	}
+
+	// Update content based on current template
+	useEffect( () => {
```

Comment:
This seems hacky to me 😄 . We use React's implementation details to update the content after the first render and by updating the `content` attribute, we also create an `undo` level in history. That's one of the reasons I had block variations in mind back then. The `php` handling remains the same, the above `content` attribute handling is done by the variations and we can provide a bit more specific info for the block (BlockCard etc..).

An alternative, if this approach is preferred, is to use `__unstableMarkNextChangeAsNotPersistent` here and a new variable for the `first content` (to avoid changing the existing `content` var).

Reply:
Haha I wasn't sure where else to turn so I jumped to `useEffect`! Thanks so much for the detailed explanation.

As soon as I moved this logic from the variations file to edit.js, I could see the advantage of using variations. However, as @scruffian has just mentioned above, I think the advantage to not using variations is that the user (or a theme) only needs to insert one block, and the block works the content out itself based on context (i.e. the template).

With that in mind, I'm going to attempt your suggested alternative. Thanks for the suggestion and input, it's really helped me understand this more.